tests/unit/lms/validation/blackboard_test.py

```diff
@@ -0,0 +1,122 @@
+                        "id": "_7851_1",
+                        "name": "File_1.pdf",
+                        "modified": "2021-06-02T15:39:51.000Z",
+                        "unknown_field": "this_should_be_included",
+                    },
+                    {
+                        "id": "_7851_2",
+                        "name": "File_2.pdf",
+                        "modified": "2021-05-02T15:39:51.000Z",
+                    },
+                    {
+                        "id": "_7851_3",
+                        "name": "File_3.pdf",
+                        "modified": "2021-04-02T15:39:51.000Z",
+                    },
+                ],
+            )
+        ],
+    )
+    def test_valid(self, input_, output):
+        schema = BlackboardListFilesSchema(
+            factories.requests.Response(json_data=input_)
+        )
+
+        result = schema.parse()
+
+        assert result == output
+
+    @pytest.mark.parametrize(
+        "json_data,messages",
+        [
+            # No top-level "results" list.
+            ({}, {"results": ["Missing data for required field."]}),
+            # The top-level "results" list is not a list.
+            ({"results": None}, {"results": ["Field may not be null."]}),
+            # A file in the "results" list is missing required fields.
+            (
+                {
+                    "results": [
+                        {
+                            "foo": "bar",
+                        }
+                    ]
+                },
+                {
+                    "results": {
+                        0: {
+                            "id": ["Missing data for required field."],
+                            "modified": ["Missing data for required field."],
+                            "name": ["Missing data for required field."],
+                        }
+                    }
+                },
+            ),
+            # A file in the "results" list has invalid values.
+            (
+                {
+                    "results": [
+                        {
+                            "id": 23,
+                            "name": None,
+                            "modified": False,
+                        },
+                    ]
+                },
+                {
+                    "results": {
+                        0: {
+                            "id": ["Not a valid string."],
+                            "modified": ["Not a valid string."],
+                            "name": ["Field may not be null."],
+                        }
+                    }
+                },
+            ),
+        ],
+    )
+    def test_invalid(self, json_data, messages):
+        schema = BlackboardListFilesSchema(
+            factories.requests.Response(json_data=json_data)
+        )
+
+        with pytest.raises(ValidationError) as exc_info:
+            schema.parse()
+
+        assert exc_info.value.messages == messages
```

Comment:
### Maybe don't test the schema messages specifically?

I don't think we actually care _why_ a particular piece of data is invalid (we just raise a blanket exception). So perhaps we should reduce the scope of these tests to include the bits we need (it raises) and exclude the parts our code doesn't actually rely on (it raises this _exact_ set of messages). This would remove the `messages` side of this and really cut this test down.

This is one potential advantage of testing the schema and the end-point together as we don't end up testing the private interface between the two. For example, if the parent function is the only consumer of the schema, and we consume all these error messages, then there's actually no point in testing they are present at all (vary this example to meet the facts etc.).

### Maybe don't check as much about the schema at all

I feel like we've had this chat before, but to what extent do we want to be testing `marshmallow` schema? It seems like we run to the point of testing the behavior of `marshmallow` pretty quick.

I'd be _strongly_ tempted to prove:

 * The schema accepts a full, validly specified document
 * The schema rejects at least a couple of obvious invalid documents

And leave it at that. Few data formats meet a schema, and infinite number don't. We've chosen a small number of the infinite that don't match. Why not check a field rejects a list as well as None, what about a list with None in it, what a list with a list with None in it... 

If every API end-point has a separate schema, and every schema has 100+ lines of test code we're signing ourselves up for thousands of lines of testing code for even a moderately sized API. All of that comes with a maintenance cost, and if it's lower value (proving that marshmallow `Str` does what you'd expect) I'm not sure there's much benefit.

The 122 lines above test 9 lines of schema code. I'd say that's not a good trade off.

Reply:
## Maybe don't test the schema messages specifically?

Done.

The reason for asserting about the schema messages was to make sure that the tests were actually testing what they thought they were testing. If the test merely asserts that `ValidationError` was raised it could be that the test set up is actually wrong and `ValidationError` is actually being raised for some other reason.

_But_ I implemented a suggestion that you made in another comment below that the tests should contain a single valid API response and then the tests for _invalid_ responses should work by taking the valid response and mutating it:

```python
# This is example code, it's not the actual code.

def test_it_raises_if_the_id_isnt_a_string(self, json_data):
    json_data["results"][0]["id"] = 23

    schema = BlackboardListFilesSchema(Response(json_data=json_data))

    with pytest.raises(ValidationError):
        schema.parse()
```

The tests know that the `json_data` fixture is a valid JSON data: there is another test above that passes `json_data` to the schema unmodified and asserts that it raises. And since this test only modified one thing about `json_data`, if `ValidationError` is raised it must have been for that one reason :+1:

Reply:
## Testing the schema and the endpoint together

I think it's better to treat schemas as separately unit-tested components. Some of the reasons are covered in other comments (e.g. consistency with request schemas and with response schemas that're used in multiple places). More generally I think it's better to break things up into small, separate components with small, separate unit tests. In the case of a validation schema unit tests that import the schema and call it directly are "closer to the metal" and as a result are much easier to understand and maintain. There's just much less going on under each test. I can look at the schema file (simple) and its unit tests (simple) and there isn't any other code involved in those unit tests, they just import the schema and call it.

Sorry to be frank here but I find Canvas API client's tests quite hard to understand and maintain. Some quite well-shrouded test bugs have been found in them and I suspect they contain more. And I suspect that if that testing style were used more widely in our codebase we'd get more of those well-camouflaged test bugs.

Reply:
## Don't just test Marshmallow

I'm sympathetic to the idea that we shouldn't write a dozen unit tests for every single schema that're really just "testing Marshmallow" (or at least testing things about the schema that're quite trivial to implement with Marshmallow and unlikely to contain any mistakes). @marcospri has also raised this point and I think he gave a good example: we don't test all the behaviours of SQLAlchemy in the unit tests for every one of our model classes.

The counter-argument is that will some schema tests might _seem_ pointless they actually do have a point: they're protecting us against changes in Marshmallow. A new Marshmallow version is released, Dependabot sends us a PR, CI passes, we assume "Must be fine then" and merge it. This is not a safe assumption if we haven't got tests for our Marshmallow schemas.

I think the solution is probably a middle-ground: do write tests for Marshmallow schemas, but just don't go crazy and write dozens of them. Which of course is what you were suggesting in your comment too.

Where we seem to differ is that I think my tests already achieve this balance. `TestBlackboardListFilesSchema` has:

* One test for a valid response
* Four tests for different kinds of invalid responses:
  1. No `"results"` list
  2. `"results"` list is not a list
  3. One of the files dicts is missing all the required fields
  4. One of the files dicts has invalid values in all the fields
 
Five simple tests total. I don't think that's excessive at all.

The tests are about 70 lines of code. About 10 lines of that is a fixture that looks to me like it might be quite generic and possibly could be shared among all schema tests (or at least all response schema tests). Another 20 lines or so is the fake Blackboard API response dict rather than actually Python code. So it's really more like 40ish lines of actual test code.